src/index.c

```diff
@@ -0,0 +1,129 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include "index.h"
+
+
+/* Call to create a new empty Index() */
+Index* init_index(int size) {
+	Index *index = (Index*) malloc(sizeof(Index));
+	if (index == NULL) {
+		printf("Failed to allocate memory for index\n");
+		return NULL;
+	}
+
+	// create contents and set all pointers to null
+	Entry **contents = (Entry **) malloc(size * sizeof(Entry*));
+	for (int i = 0; i < size; i++) {
+		*(contents + i) = NULL;
+	}
+
+	if (contents == NULL) {
+		printf("Failed to allocate memory for index contents.\n");
+		free(index);
+		return NULL;
+	}
+
+	index->size = size;
+	index->positions_filled = 0;
+	index->contents = contents;
+	return index;
+}
+
+/* Inserts a new entry into hash table; where there are
+ * collisions, this program uses chaining. */
+int index_insert(Index *index, int key, char *value) {
+
+	int position = hash(key, index->size);
+
+	Entry *new_entry = (Entry*) malloc(sizeof(Entry));
+	if (new_entry == NULL) {
+		printf("Failed to allocate memory for new hash table entry.\n");
+		return -1;
+	}
+	new_entry->key = key;
+	new_entry->value = value;
+	new_entry->next = NULL;
+
+	// see if the key is already in the hash map
+	Entry *in_slot = *(index->contents + position);
+	Entry *trav = in_slot;
+	if (trav) {
+		while (trav) {
+			if (trav->key == key) {
+				trav->value = value;
+				return 0;
```

Comment:
In this case, do we need to free the memory we allocated above on line 39? Maybe you could find the location first and only then allocate memory if needed. If we already have an `Entry` for the key, it looks like we can just re-use the allocated `Entry`

Reply:
ya make sense, doing that now.

Reply:
Thanks.